### PR TITLE
[WIP] Adding Jaccard index for computing atlas similarity

### PR DESCRIPTION
**Pull request by Joaoloula**
Work needed on examples and documentation. We could also consider adding other measures of similarity (ICC, mutual information).

### CODE REVIEW

**Comment by GaelVaroquaux** [comment]
Jaccard similarity is already in scikit-learn:
http://scikit-learn.org/stable/modules/generated/sklearn.metrics.jaccard_similarity_score.html

Overall, this kind of functionality, not strictly neuroimaging related, lives more in scikit-learn than nilearn.

What may be useful could be to add a function to do matching of 2 sets of maps, using sklearn.utils.linear_assignment. It should also be added to a nilearn example, for instance at the end of http://nilearn.github.io/auto_examples/03_connectivity/plot_compare_resting_state_decomposition.html, to plot the same map for both ICA and dict-learning.

**Comment by AlexandreAbraham** [comment]
> Jaccard similarity is already in scikit-learn:
http://scikit-learn.org/stable/modules/generated/sklearn.metrics.jaccard_similarity_score.html

Scikit-learn implements the Jaccard Index in the specific setting of measuring clustering similarity. Here we have a fuzzy extension of the metric that will probably not be accepted in sklearn because there is no "best" way to make this extension. In particular, people in neuroimaging tend to use extensions that do not rely on linear assignment of the regions, see _Generalized Overlap Measures for Evaluation and
Validation in Medical Image Analysis_ (Crum 2006)

**Comment by GaelVaroquaux** [comment]
I won't accept this in nilearn unless it comes with clear neuroimaging
applications. Everything in nilearn should solve a brain-research
problem.

What's the problem that you are trying to solve?

**Comment by AlexandreAbraham** [comment]
> What's the problem that you are trying to solve?

Having a similarity score between two sets of brain maps with continuous values and possible overlaps.

**Comment by GaelVaroquaux** [comment]
>     What's the problem that you are trying to solve?

> Having a similarity score between two sets of brain maps with
> continuous values and possible overlaps.

No, that's a formal problem. In what application setting? What the
'brain' question?

**Comment by Joaoloula** [comment]
I'd argue one of the main interests is a matter of reproducibility; we could introduce an example in which we split a dataset and build atlases from each of the splits and compare their similarity (the brain question being: are the regions we're learning serving actual stable and interesting neuroscientific purposes or are they just an artifact of the specific subset we looked at?)

**Comment by GaelVaroquaux** [comment]
You don't have to convince me it is useful, I am convinced. You have to
put it in a way that non technical users can use it, understand how and
why.

Nilearn targets people who are not computer scientists. If we cannot
demonstrate how some piece of code makes their life better, it does not
enter nilearn.

**Comment by AlexandreAbraham** [comment]
So! People here wants to assess the _reliability_ of methods extracting resting state networks. They define _reliability_ as the ability to extract similar brain atlases on test-retest sessions. In particular, most of them want to compare ICA to their approach. This is what I understood!

**Comment by GaelVaroquaux** [comment]
Yes, but just giving them a metric won't help them solve their problem.
They need more. It's like giving people a showel when they want to build
a house.

Once again: either you can do a convincing example, or it does not go
into nilearn. Without a convincing example, it either won't be used, or
we'll keep getting questions on how to use it.

That's a clear and important rule of nilearn: anything that goes in must
produce an example that is relevant for brain imaging, even for people
who are not computer scientists.

**Comment by Joaoloula** [comment]
I think the following example is quite didactic: in the spirit of 8.4.7, we take ADHD and we split it in two. We then create atlases for both splits using CanICA and Dictionary Learning. Finally, we compute two Jaccard indexes: one between the CanICA atlases and one between the DictLearning atlases, to compare replicability between methods. The narrative would explain that we want our methods to be not only valid, but also replicable, and would give an intuitive explanation of the index.

**Comment by Joaoloula** [comment]
A rough idea of the example: http://bit.ly/2cVX1Hh

**Comment by GaelVaroquaux** [comment]
OK. Make sure that the example runs fast enough. Maybe limit yourself to
6 subjects, and make 2 different sub-groups of 5 subjects, with only 10
components.

To make an example that is easy to follow, I think that you need
something higher level than just the Jaccard index. It's too low level,
and will lead to a lot of numpy code in the example. Maybe a function
that take 2 set of maps, returns a reordered version and a list of
similarities.

**Comment by Joaoloula** [comment]
Simplified example (with 10 subjects because the atlas using 6 subjects were just too ugly): http://bit.ly/2cVX1Hh. I think it's reasonably high-level and light on numpy code (most of the example is just *getting* the atlases, which is quite a standard analysis, and computing the jaccard indexes is done in one line).

I really like the idea of comparing two atlases map by map and trying to find a one-to-one correspondence, and I believe a lot of people would be interested in it. However, it seems to me like a significantly different question than just measuring similarity between atlases, shouldn't we leave it to another PR?

**Comment by GaelVaroquaux** [comment]
> Merged #1284.

Not happy

**Comment by AlexandreAbraham** [comment]
Sorry for that. I merge #1316 but I did not see that it contained this commit. I reverted the PR.

**Comment by AlexandreAbraham** [comment]
OK, I figures out what happened. @Joaoloula made a PR where this commit was pushed and reverted. I did not noticed that when merging (damn merging from phone). So this PR is considered as merged but the code never ended up in the repo.

Joao, could you make two clean PRs for #1316 and this one?

**Comment by Joaoloula** [comment]
Sorry about that, totally my fault: I got mixed up with the working branch for the other PR and it ended up pushing this one, since the changes had been reverted I didn't think it would change anything. I didn't know this would get automatically merged as well. I'll create new, clean PRs.
